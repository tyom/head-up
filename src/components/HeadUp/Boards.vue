<template>
  <div class="HeadUpBoards flex-grow overflow-hidden">
    <Board
      v-for="(board, idx) in boards"
      v-bind="board"
      :key="board.title + board.id + idx"
    />
    <slot />
  </div>
</template>

<script>
import Board from '../Board';

export default {
  components: {
    Board,
  },
  props: {
    boards: {
      type: Array,
      default: () => [],
    },
  },
};
</script>
