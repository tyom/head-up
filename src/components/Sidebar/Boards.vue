<template>
  <div class="Boards">
    <transition-group tag="ul" name="boardReveal" class="board-list">
      <li
        v-for="board in boards"
        :key="board.id"
        :class="getBoardListItemClass(board)"
        :title="getBoardTitle(board)"
        class="relative"
      >
        <transition name="actionsReveal">
          <div v-if="editMode && board.editable" class="board-actions">
            <button
              type="button"
              class="remove-button"
              @click="$emit('remove', board.id)"
            >
              <v-icon name="minus-circle" />
            </button>
          </div>
        </transition>
        <button class="board-button" @click="$emit('activate', board.id)">
          <Board :id="board.id" :cells="board.cells" preview class="board" />
        </button>
        <div v-if="board.title" class="board-title">
          {{ board.title }}
        </div>
      </li>
    </transition-group>
  </div>
</template>

<script>
import Board from '../Board';

export default {
  components: {
    Board,
  },
  props: {
    boards: {
      type: Array,
      default: () => [],
    },
    activeId: {
      type: String,
      default: '',
    },
    editMode: {
      type: Boolean,
      default: false,
    },
  },
  methods: {
    getBoardTitle(board) {
      return board.title + (board.editable ? '' : ' (read-only)');
    },
    getBoardListItemClass(board) {
      return {
        _active: this.activeId === board.id,
        ['_read-only']: !board.editable,
      };
    },
  },
};
</script>

<style scoped>
.board-list {
  @apply grid gap-3;
}

.board-title {
  @apply text-xs text-center mt-1 -mx-2 transition duration-100;
}

.board {
  @apply transition duration-200 relative border;
  border-color: #fff4;
  min-width: 4rem;
  height: 10vh;

  &:hover {
    border-color: #fff8;
  }

  @nest ._active &,
  &:active {
    @apply border-white;
  }
}

.board-button {
  @apply block w-full p-px outline-none cursor-default;

  &.focus-visible {
    @apply shadow-outline;
  }
}

.board:hover + .board-title {
  @apply text-white;
}

._read-only .board::after {
  @apply opacity-0 absolute inset-0 transition duration-200 pointer-events-none;
  content: '';
  background: repeating-linear-gradient(
    -45deg,
    #fff0,
    #fff0 0.5rem,
    #fff1 0.5rem,
    #fff1 1rem
  );

  @nest ._edit & {
    @apply opacity-100;
  }
}

.board-actions {
  @apply absolute z-10;
  top: -0.2rem;
  right: -0.2rem;
}

.remove-button {
  @apply flex p-px bg-white rounded-full outline-none text-red-800 shadow;

  & .fa-icon {
    @apply w-3 h-3;
  }

  &:hover {
    @apply text-red-700;
  }

  &:active {
    @apply text-red-900;
  }
}

.boardReveal-enter-active,
.boardReveal-leave-active {
  transition: 0.1s;
}
.boardReveal-enter,
.boardReveal-leave-to {
  opacity: 0;
  transform: translateY(-100%);
}

.actionsReveal-enter-active,
.actionsReveal-leave-active {
  transition: 0.1s;
}
.actionsReveal-enter,
.actionsReveal-leave-to {
  transform: scale(0.1);
}
</style>
