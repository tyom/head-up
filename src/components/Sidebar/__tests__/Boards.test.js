import { shallowMount } from '@vue/test-utils';
import Boards from '../Boards';

const boardsMock = [
  {
    id: '1',
    title: 'Board #1',
    cells: [{ title: 'Cell #1' }, { title: 'Cell #2' }],
  },
  {
    id: '2',
    title: 'Board #2',
    editable: true,
    cells: [{ title: 'Cell #1' }, { title: 'Cell #2' }],
  },
];

test('render default', () => {
  const wrapper = shallowMount(Boards);
  expect(wrapper).toMatchSnapshot();
});

test('render with boards', () => {
  const wrapper = shallowMount(Boards, {
    propsData: {
      boards: boardsMock,
    },
  });

  expect(wrapper).toMatchSnapshot();
});

test('render with active board', () => {
  const wrapper = shallowMount(Boards, {
    propsData: {
      boards: boardsMock,
      activeId: '1',
    },
  });

  expect(wrapper).toMatchSnapshot();
});

test('render in edit mode', () => {
  const wrapper = shallowMount(Boards, {
    propsData: {
      boards: boardsMock,
      editMode: true,
    },
  });

  expect(wrapper).toMatchSnapshot();
});

test('respond to board removal', () => {
  const wrapper = shallowMount(Boards, {
    propsData: {
      boards: boardsMock,
      editMode: true,
    },
  });

  const removeButton = wrapper.find('.remove-button');
  removeButton.trigger('click');
  expect(wrapper.emitted('remove')[0]).toEqual(['2']);
});

test('respond to board activation', () => {
  const wrapper = shallowMount(Boards, {
    propsData: {
      boards: boardsMock,
      activeId: '1',
    },
  });

  const boards = wrapper.findAll('.board');
  boards.at(1).trigger('click');
  expect(wrapper.emitted('activate')[0]).toEqual(['2']);
  boards.at(0).trigger('click');
  expect(wrapper.emitted('activate')[1]).toEqual(['1']);
});
