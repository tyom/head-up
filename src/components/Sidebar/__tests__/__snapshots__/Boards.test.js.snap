// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render default 1`] = `
<div class="Boards">
  <transition-group-stub tag="ul" name="boardReveal" class="board-list"></transition-group-stub>
</div>
`;

exports[`render in edit mode 1`] = `
<div class="Boards">
  <transition-group-stub tag="ul" name="boardReveal" class="board-list">
    <li title="Board #1 (read-only)" class="relative _read-only">
      <transition-stub name="actionsReveal">
        <!---->
      </transition-stub> <button class="board-button">
        <board-stub id="1" isthumb="true" title="" cells="[object Object],[object Object]" class="board"></board-stub>
      </button>
      <div class="board-title">
        Board #1
      </div>
    </li>
    <li title="Board #2" class="relative">
      <transition-stub name="actionsReveal">
        <div class="board-actions"><button type="button" class="remove-button">
            <v-icon name="minus-circle"></v-icon>
          </button></div>
      </transition-stub> <button class="board-button">
        <board-stub id="2" isthumb="true" title="" cells="[object Object],[object Object]" class="board"></board-stub>
      </button>
      <div class="board-title">
        Board #2
      </div>
    </li>
  </transition-group-stub>
</div>
`;

exports[`render with active board 1`] = `
<div class="Boards">
  <transition-group-stub tag="ul" name="boardReveal" class="board-list">
    <li title="Board #1 (read-only)" class="relative _active _read-only">
      <transition-stub name="actionsReveal">
        <!---->
      </transition-stub> <button class="board-button">
        <board-stub id="1" isthumb="true" title="" cells="[object Object],[object Object]" class="board"></board-stub>
      </button>
      <div class="board-title">
        Board #1
      </div>
    </li>
    <li title="Board #2" class="relative">
      <transition-stub name="actionsReveal">
        <!---->
      </transition-stub> <button class="board-button">
        <board-stub id="2" isthumb="true" title="" cells="[object Object],[object Object]" class="board"></board-stub>
      </button>
      <div class="board-title">
        Board #2
      </div>
    </li>
  </transition-group-stub>
</div>
`;

exports[`render with boards 1`] = `
<div class="Boards">
  <transition-group-stub tag="ul" name="boardReveal" class="board-list">
    <li title="Board #1 (read-only)" class="relative _read-only">
      <transition-stub name="actionsReveal">
        <!---->
      </transition-stub> <button class="board-button">
        <board-stub id="1" isthumb="true" title="" cells="[object Object],[object Object]" class="board"></board-stub>
      </button>
      <div class="board-title">
        Board #1
      </div>
    </li>
    <li title="Board #2" class="relative">
      <transition-stub name="actionsReveal">
        <!---->
      </transition-stub> <button class="board-button">
        <board-stub id="2" isthumb="true" title="" cells="[object Object],[object Object]" class="board"></board-stub>
      </button>
      <div class="board-title">
        Board #2
      </div>
    </li>
  </transition-group-stub>
</div>
`;
